application/apps/indexer/session/src/state/indexes/keys.rs

```diff
@@ -0,0 +1,133 @@
+use crate::events::{NativeError, NativeErrorKind};
+use indexer_base::progress::Severity;
+
+#[derive(Debug)]
+pub struct Keys {
+    keys: Vec<u64>,
+    sorted: bool,
+}
+
+impl Keys {
+    pub fn new() -> Self {
+        Keys {
+            keys: vec![],
+            sorted: false,
+        }
+    }
+    pub fn add(&mut self, position: u64) {
+        self.keys.push(position);
+        self.sorted = false;
+    }
+
+    pub fn remove(&mut self, position: &u64) {
+        if let Ok(index) = self.keys.binary_search(position) {
+            self.keys.remove(index);
```

Comment:
not sure why this should work...`binary_search` only works on sorted vector...when calling remove, the vector is not necessarily sorted.

Reply:
and `map` (on level upper) keeps `keys` always stored. That's why `Map` works fast - we are sorting as soon as we need it.